pkg/tests/end_to_end_tests/ha_multiple_promscales_test.go

```diff
@@ -0,0 +1,238 @@
+						leaseStart: time.Unix(0, 0),
+						leaseUntil: time.Unix(100, 0),
+					},
+				},
+				tickSyncRoutine: true,
+			},
+		}}
+	runHATest(t, testCase)
+}
+
+func TestHALeaderChangedFromOutside(t *testing.T) {
+	testCase := haTestCase{
+		db: "ha_leader_change_from_outside",
+		steps: []haTestCaseStep{
+			{
+				desc: "1 becomes leader; don't run sync; leader is changed from outside",
+				input: haTestInput{
+					replica: "1",
+					minT:    time.Unix(0, 0),
+					maxT:    time.Unix(0, 0),
+				},
+				output: haTestOutput{
+					expectedNumRowsInDb: 1,
+					expectedMaxTimeInDb: time.Unix(0, 0),
+					expectedLeaseStateInDb: leaseState{
+						cluster:    "cluster",
+						leader:     "1",
+						leaseStart: time.Unix(0, 0),
+						leaseUntil: time.Unix(60, 0),
+					},
+				},
+				tickSyncRoutine: false,
+				explicitLease: &leaseState{
+					cluster:    "cluster",
+					leader:     "2",
+					leaseStart: time.Unix(60, 0),
+					leaseUntil: time.Unix(120, 0),
+				},
+			}, {
+				desc: "1 tries to send data in cached lease, all is good; sync",
+				input: haTestInput{
+					replica: "1",
+					minT:    time.Unix(1, 0),
+					maxT:    time.Unix(1, 0),
+				},
+				output: haTestOutput{
+					expectedNumRowsInDb: 2,
+					expectedMaxTimeInDb: time.Unix(1, 0),
+					expectedLeaseStateInDb: leaseState{
+						cluster:    "cluster",
+						leader:     "2",
+						leaseStart: time.Unix(60, 0),
+						leaseUntil: time.Unix(120, 0),
+					},
+				},
+				tickSyncRoutine: true,
+			}, {
+				desc: "1 is no longer leader, can't send data",
+				input: haTestInput{
+					replica: "1",
+					minT:    time.Unix(2, 0),
+					maxT:    time.Unix(2, 0),
+				},
+				output: haTestOutput{
+					expectedNumRowsInDb: 2,
+					expectedMaxTimeInDb: time.Unix(1, 0),
+					expectedLeaseStateInDb: leaseState{
+						cluster:    "cluster",
+						leader:     "2",
+						leaseStart: time.Unix(60, 0),
+						leaseUntil: time.Unix(120, 0),
+					},
+				},
+				tickSyncRoutine: true,
+			},
+		}}
+	runHATest(t, testCase)
+}
+
+func TestHALeaderChangedFromOutsideCacheNotAwareAtFirst(t *testing.T) {
+	testCase := haTestCase{
+		db: "ha_leader_change_from_outside",
+		steps: []haTestCaseStep{
+			{
+				desc: "1 becomes leader; don't run sync; leader is changed from outside",
+				input: haTestInput{
+					replica: "1",
+					minT:    time.Unix(0, 0),
+					maxT:    time.Unix(0, 0),
+				},
+				output: haTestOutput{
+					expectedNumRowsInDb: 1,
+					expectedMaxTimeInDb: time.Unix(0, 0),
+					expectedLeaseStateInDb: leaseState{
+						cluster:    "cluster",
+						leader:     "1",
+						leaseStart: time.Unix(0, 0),
+						leaseUntil: time.Unix(60, 0),
+					},
+				},
+				tickSyncRoutine: false,
+				explicitLease: &leaseState{
+					cluster:    "cluster",
+					leader:     "2",
+					leaseStart: time.Unix(60, 0),
+					leaseUntil: time.Unix(120, 0),
+				},
+			}, {
+				desc: "2 tries to send data in lease, lease is cached from before-> rejected; sync",
```

Comment:
this is the same less than Ideal case here. I guess we should try to do something clever if the input is for a point past the cached leaseUntil. Let's think about this case a bit more

Reply:
You mean to try to sync with the db if we have data after leaseUntil from a non-leader instance?